gr-analog/lib/agc3_cc_impl.cc

```diff
@@ -56,81 +56,137 @@ agc3_cc_impl::agc3_cc_impl(float attack_rate,
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_gain;
+}
+float agc3_cc_impl::max_gain() const
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    return d_max_gain;
+}
+
+void agc3_cc_impl::set_attack_rate(float rate)
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_attack = rate;
+}
+void agc3_cc_impl::set_decay_rate(float rate)
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_decay = rate;
+}
+void agc3_cc_impl::set_reference(float reference)
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_reference = reference;
+}
+void agc3_cc_impl::set_gain(float gain)
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_gain = gain;
+    d_gain_change = true;
+}
+void agc3_cc_impl::set_max_gain(float max_gain)
+{
+    gr::thread::scoped_lock guard(d_setter_mutex);
+    d_max_gain = max_gain;
+}
+
 int agc3_cc_impl::work(int noutput_items,
                        gr_vector_const_void_star& input_items,
                        gr_vector_void_star& output_items)
 {
-    const gr_complex* in = (const gr_complex*)input_items[0];
-    gr_complex* out = (gr_complex*)output_items[0];
-
-#ifdef __GNUC__
-    // Compute a linear average on reset (no expected)
-    if (__builtin_expect(d_reset, false)) {
-        float mags[noutput_items] __attribute__((aligned(16)));
-        volk_32fc_magnitude_32f(mags, &in[0], noutput_items);
-#else
-    // Compute a linear average on reset (no expected)
-    if (!d_reset) {
-        _declspec(align(16)) std::vector<float> mags(noutput_items);
-        volk_32fc_magnitude_32f(&mags[0], &in[0], noutput_items);
-#endif
-        float mag(0.0);
-        for (int i = 0; i < noutput_items; i++) {
-            mag += mags[i];
+    auto in = reinterpret_cast<const gr_complex*>(input_items[0]);
+    auto out = reinterpret_cast<gr_complex*>(output_items[0]);
+    float attack;
+    float decay;
+    float reference;
+    float gain;
+    float max_gain;
+    unsigned int iir_update_decim;
+    // make copies of the runtime-adjustable floats
+    // we do this so we can be finely granular when locking
+    {
+        gr::thread::scoped_lock guard(d_setter_mutex);
+        attack = d_attack;
+        decay = d_decay;
+        reference = d_reference;
+        gain = d_gain;
+        max_gain = d_max_gain;
+        iir_update_decim = d_iir_update_decim;
+        d_gain_change = false; // if gain setting happened, we've just used the new gain.
+    }
+    if (d_reset) {
+        float magnitude = 0.0f;
+        for (unsigned int idx = 0; idx < iir_update_decim * 4; ++idx) {
+            magnitude += std::abs(in[idx]);
         }
-        d_gain = d_reference * (noutput_items / mag);
+        gain = reference * (iir_update_decim * 4 / magnitude);
 
-        if (d_gain < 0.0)
-            d_gain = 10e-5;
-
-        if (d_max_gain > 0.0 && d_gain > d_max_gain) {
-            d_gain = d_max_gain;
-        }
+        std::clamp(gain, 1e-4f, max_gain);
```

Comment:
Switching to `std::clamp` seems like it's changing the behaviour. For instance, the documentation says that `max_gain = 0` means unlimited, but that will no longer be the case. Also, the previous version only set the gain to `10e-5` in the event that `d_gain` was negative. (I'm not even sure how that could happen, unless `reference` was negative!)

Reply:
The fact that this change didn't break any tests also makes it clear that the `max_gain = 0` behaviour is untested.

Reply:
I'm also not really sure the default max_gain makes any sense, 2¹⁶ seems very arbitrary.

Reply:
also, the setting of a gain to 1e-4 in case the reference was negative, um, let's simply get rid of that: 1e-4 might be an enormous gain in some settings, and we'd have the most funky behaviour. A negative reference (power!) makes no sense, so we reject it in the setter.